src/RTI_XMLOutputUtility.c

```diff
@@ -0,0 +1,220 @@
+    DDS_DomainParticipantFactory *factory = DDS_TheParticipantFactory;
+    struct DDS_XMLObject *root = NULL, *default_profile = NULL;
+    struct DDS_DomainParticipantFactoryQos dpf_qos = DDS_DomainParticipantFactoryQos_INITIALIZER;
+    struct CommandLineArguments cmd_args;
+    char *user_qos_profile_file = NULL;
+    RTI_Retval result = RTI_ERROR;
+
+    CommandLineArguments_initialize(&cmd_args);
+    result = CommandLineArgumentParser_parse_arguments(argc, argv, &cmd_args);
+    CommandLineArgumentParser_print_arguments(&cmd_args);
+
+    if (result == RTI_ERROR) {
+        goto done;
+    }
+
+    if (cmd_args.user_file != NULL) {
+        if (Common_allocate_string(
+                    &user_qos_profile_file, 
+                    strlen("file://") + strlen(cmd_args.user_file)) != RTI_OK) {
+            printf("String allocation for file name '%s' failed! \n", 
+                    cmd_args.user_file);
+            goto done;
+        }
+        sprintf(
+                user_qos_profile_file, 
+                "file://%s", 
+                cmd_args.user_file);
+
+        DDS_StringSeq_ensure_length(&dpf_qos.profile.url_profile, 1, 1);
+        *DDS_StringSeq_get_reference(&dpf_qos.profile.url_profile, 0) 
+                = DDS_String_dup(user_qos_profile_file);
+        DDS_DomainParticipantFactory_set_qos(factory, &dpf_qos);
+    }
+
+    xml_save_context.depth = 1;
+    if (DDS_DomainParticipantFactory_load_profiles(factory) 
+            != DDS_RETCODE_OK) {
+        printf("Failed to load the XML configuration files \n");
+        goto done;
+    }
+
+    if (cmd_args.qos_library == NULL || cmd_args.qos_profile == NULL) {
+        root = DDS_QosProvider_get_xml_root(
+                DDS_DomainParticipantFactory_get_qos_providerI(factory));
+        if (root != NULL) {
+            default_profile = (struct DDS_XMLObject *) DDS_XMLDds_get_default_qos_profile(
+                    (struct DDS_XMLDds *) root);
+            if (default_profile != NULL) {
+                cmd_args.qos_library = 
+                        DDS_String_dup(
+                                DDS_XMLObject_get_name(
+                                        DDS_XMLObject_get_parent(default_profile)));
+                cmd_args.qos_profile = 
+                        DDS_String_dup(
+                                DDS_XMLObject_get_name(default_profile));
+                printf("The default <qos_profile> detected was '%s::%s'! \n", 
+                        cmd_args.qos_library, 
+                        cmd_args.qos_profile);
+            } else {
+                printf("There is no <qos_profile> marked with is_default_qos=\"true\", "
+                        "getting the default QoS values for <%s> now! \n", 
+                        cmd_args.qos_type);
+            }
+        } else {
+            printf("Failed to get the root of the DOM tree! \n");
+            goto done;
+        }
+    }
+
+    while (1) {
```

Comment:
Yes we need that because the first pass will tell us the size of the string we need to allocate. And the second pass will actually decide where to output the values.

Reply:
First I don't think (and we should not) we need to parse the command twice. Also a while(1) is not correct semantics here. You want to do it with a while (length ==0 || ptr == NULL). We do that in other places in the product and we don't use while(1).

Reply:
Done. I got rid of the loop and added some comments to make things clear.